utils/das_data_sorter.py

```diff
@@ -1,76 +1,170 @@
+        '''
+        if args.unit == "seconds" or args.unit == "s":
+            milliseconds = milliseconds/1000
+        elif args.unit == "minutes" or args.unit == "m":
+            milliseconds = milliseconds/1000/60
+        else:
+            raise ValueError("Argument only accepts either seconds/s or minutes/m")
+        
+        # sets the groups of indexes to use for averaging in future
+        self.indexes = self.group_index(milliseconds)
+
+        return range(1,len(self.indexes)+1)
+
+    def group_index(self, time:D) -> D:
+        '''Returns a universal array of arrays of indexes, based on the specified time unit. Each array represents the indexes 
+        of the data points within the same time interval. (eg. 1123ms and 1748ms are in the same time interval for seconds, 
+        but 2453ms isn't)
+
+        When given a valid value for args.unit, function will group the indexes within the same time interval in an array.
+        Once all the indexes of a time interval has been accounted for, the array of indexes will be pushed into a
+        universal array so that each element represents the data points of one singular time interval.
+        '''
+        universal_array = []
+        index_array = []
+        previous_time = 0
+
+        for i in range(len(time)):
+            if time[i] > previous_time:
+                # if true, then time at time[i] is the next second/minute
+                if previous_time != 0:
+                    # appends index array into universal array, but ignores the first iteration
+                    universal_array.append(index_array)
+
+                previous_time = ceil(time[i])
+                index_array = [] # recreates new array if previous time has changed
+            index_array.append(i) 
+
+        # pushes the final array when for loop has completed
+        universal_array.append(index_array)
+
+        return universal_array
+
+    def mean(self, data_array:list) -> int:
+        '''Finds the average of a given set of numbers. 
+
+        Disregards None and zero in calculation. Otherwise, if input is invalid, then raises error.
+        '''
+        total = 0
+        length = len(data_array)
+        for number in data_array:
+            if number == 0 or pd.isna(number):
+                # None and zeroes are ignored
+                length -= 1
+                continue
+            elif number > 0 or number < 0:
+                # 'number' value is valid
+                total += number
+            else:
+                # 'number' value is invalid
+                raise ValueError("Data point invalid and is neither zero nor None. The data point was "+str(number)+". ("+str(type(number))+")")
+        
+        try:
+            return total/length
+        except ZeroDivisionError:
+            # in the event where length = 0, due to all the elements in data_array being ignored
+            return 0 
+
+    def average_data(self, data:D) -> D:
```

Comment:
I just realised it looks like theres a smiley face and a shocked face `:D) -> D:`

Reply:
The function takes a happy face and makes it sad :(

Reply:
@hallgchris why is it `average_data`? Why can't it be `happy_data`?